school/database.py

```diff
@@ -0,0 +1,588 @@
+
+        if degree:
+            degree = degree.upper()
+            query = query.filter(Program.degree.like(f"{degree}%"))
+
+        return query.all()
+
+    def get_or_create(ctx, abbreviation: str, degree: str) -> Program:
+        query = (
+            session.query(Program)
+            .filter_by(guild_id=ctx.guild.id, abbreviation=abbreviation, degree=degree)
+            .one_or_none()
+        )
+
+        if query:
+            return query
+
+        program = Program(
+            guild_id=ctx.guild.id, abbreviation=abbreviation, degree=degree
+        )
+
+        session.add(program)
+        session.commit()
+
+        return program
+
+    def delete(self):
+        session.delete(self)
+        session.commit()
+
+    def __repr__(self) -> str:
+        return (
+            f'<{self.__class__.__name__} idx="{self.idx}" guild_id="{self.guild_id}" '
+            f'abbreviation="{self.abbreviation}" name="{self.name}" degree="{self.degree}">'
+        )
+
+    def dump(self) -> dict:
+        return {
+            "idx": self.id,
+            "guild_id": self.guild_id,
+            "abbreviation": self.abbreviation,
+            "name": self.name,
+            "degree": self.degree,
+            "subjects": self.subjects,
+        }
+
+
+class SubjectUrl(database.base):
+    __tablename__ = "school_dataset_subject_url"
+
+    subject_id = Column(
+        Integer,
+        ForeignKey("school_dataset_subject.idx", ondelete="CASCADE"),
+        primary_key=True,
+    )
+    url = Column(String, primary_key=True)
+    subject = relationship("Subject", back_populates="url")
+
+    def __repr__(self) -> str:
+        return f'<{self.__class__.__name__} subject_ixd="{self.subject_ixd}" url="{self.url}">'
+
+    def dump(self) -> dict:
+        return {
+            "subject_ixd": self.subject_ixd,
+            "url": self.url,
+        }
+
+
+class Subject(database.base):
+    __tablename__ = "school_dataset_subject"
+    __table_args__ = (
+        UniqueConstraint(
+            "abbreviation", "guild_id", name="abbreviation_guild_id_unique"
+        ),
+    )
+
+    idx = Column(Integer, primary_key=True, autoincrement=True)
+    guild_id = Column(BigInteger)
+    abbreviation = Column(String)
+    name = Column(String)
+    institute = Column(String)
+    semester = Column(String)
+    guarantor_id = Column(
+        Integer, ForeignKey("school_dataset_teacher.idx", ondelete="SET NULL")
+    )
+    guarantor = relationship(
+        lambda: Teacher,
+        back_populates="guaranted_subjects",
+    )
+    teachers = relationship(
+        lambda: Teacher,
+        secondary=teachers_subjects,
+        back_populates="subjects",
+    )
+    programs = relationship(
+        "SubjectProgram", back_populates="subject", cascade="all, delete-orphan, delete"
+    )
+    url = relationship(
+        "SubjectUrl", back_populates="subject", cascade="all, delete-orphan, delete"
+    )
```

Comment:
Since all this is in the same repository, why don't you make a relationship with a Review here? So that you can quickly and easily see whether the subject has any or not. This will make the delete process easier.

Reply:
Yes, it does, but since School can provide data for more than one repo, I'd like to keep this separated.

Reply:
I'd like a neutral 3rd party here, @sinus-x, @BorekZnovustvoritel ?